vividus-plugin-web-app/src/main/java/org/vividus/ui/web/configuration/WebApplicationConfiguration.java

```diff
@@ -29,11 +34,18 @@
     private int tabletScreenResolutionWidthThreshold;
     private String applicationEnvironmentType;
     private final ThreadLocal<ConfigurationStorage> configurationStorage;
+    private final Supplier<URI> mainApplicationPageUriSupplier;
 
     public WebApplicationConfiguration(String mainApplicationPageUrlString, AuthenticationMode authenticationMode)
     {
         configurationStorage = ThreadLocal.withInitial(() ->
             new ConfigurationStorage(mainApplicationPageUrlString, authenticationMode));
+        mainApplicationPageUriSupplier = Suppliers.memoize(() ->
+        {
+            URI mainApplicationPageUrl = configurationStorage.get().getMainApplicationPageUri();
+            isTrue(mainApplicationPageUrl != null, "URL of the main application page should be non-blank");
+            return mainApplicationPageUrl;
+        });
```

Comment:
it won't work in this way, because Supplier is shared across threads now, but configuration storage is not, it may lead to wrong URL returned from the cache